ast/format.go

```diff
@@ -0,0 +1,356 @@
+package ast
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Returns a valid query for a given AST rooted at node `n`.
+func Format(n Node) string {
+	f := &formatter{new(strings.Builder)}
+	f.formatNode(n)
+	return f.get()
+}
+
+type formatter struct {
+	*strings.Builder
+}
+
+func (f *formatter) get() string {
+	return f.String()
+}
+
+// strings.Builder's methods never returns a non-nil error.
+func (f *formatter) writeString(s string) {
+	_, err := f.WriteString(s)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (f *formatter) writeRune(r rune) {
+	_, err := f.WriteRune(r)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (f *formatter) formatProgram(n *Program) {
+	sep := '\n'
+	for i, c := range n.Body {
+		f.formatNode(c)
+		if i < len(n.Body)-1 {
```

Comment:
just a nit so not important to change but I find this pattern easier to read.

```
for i, c := range n.Body {
    if i != 0 {
         f.writeRune(sep)
    }
    f.formatNode(c)
}
```

Reply:
much better, absolutely